### PR TITLE
Finish pile_map support enough for a basic transformer example

### PR DESCRIPTION
**Pull request by axch**
The remaining operations to cover were
- Slicing a non-ragged axis when ragged axes are present
- Einsum where both tensor and contraction dimensions are ragged

Also improved the error message that happens when trying to broadcast a ragged axis to a rectangular one

### CODE REVIEW

File: tests/dynamic_api_test.py [hunk 1]
```diff
@@ -1627,7 +1639,7 @@ def func(size):
     data = jax.lax.broadcasted_iota('int32', (3, 5, 7), 1)
     self.assertAllClose(p.data, data)
 
-  def test_ragged_einsum(self):
+  def test_einsum_ragged_tensor(self):
```
Comment (mattjj) [inline tests/dynamic_api_test.py:1642]:
```suggestion
  def test_einsum_ragged_array(self):
```

😃

File: tests/dynamic_api_test.py [hunk 2]
```diff
@@ -1640,6 +1652,31 @@ def fprop_layer(x_size):
     self.assertRegex(str(p.aval), r'Var[0-9]+:3 => i32\[3,bint\{≤5\}\[3\] with value: \[3 1 4\]\.Var[0-9]+,2,7\]')
     self.assertEqual(p.data.shape, (3, 3, 5, 2, 7))
 
+  def test_einsum_ragged_tensor_and_contract(self):
```
Comment (mattjj) [inline tests/dynamic_api_test.py:1655]:
```suggestion
  def test_einsum_ragged_array_and_contract(self):
```

😃

File: jax/_src/lax/slicing.py [hunk 3]
```diff
@@ -1287,10 +1295,17 @@ def _gather_batching_rule(batched_args, batch_dims, *, dimension_numbers,
         offset_dims=offset_dims,
         collapsed_slice_dims=collapsed_slice_dims,
         start_index_map=start_index_map)
-    return gather(operand, indices, dimension_numbers=dnums,
-                  slice_sizes=slice_sizes, unique_indices=unique_indices,
+    # TODO(reviewer): This should be correct for `operand_bdim` being a
+    # `RaggedAxis` as long as we are not gathering from any ragged axis, and as
+    # long as we have no collapsed_slice_dims.  (The latter could require
+    # adjusting the `ragged_axes` field of `operand_bdim`).  Should we put a
+    # check to confirm?
```
Comment (mattjj) [inline jax/_src/lax/slicing.py:1302]: Let's definitely put a check! Any reason not to? (We can raise NotImplemented in unhandled cases iiuc).

**Review by mattjj — APPROVED** [review]
Nice!

File: tests/dynamic_api_test.py (hunk 1 above)
Comment (axch) [reply tests/dynamic_api_test.py:1642]: I meant that the tensor dimension is ragged, in both cases.

File: jax/_src/lax/slicing.py (hunk 3 above)
Comment (axch) [reply jax/_src/lax/slicing.py]: Shape prediction improved and check added.

File: tests/dynamic_api_test.py (hunk 1 above)
Comment (axch) [reply tests/dynamic_api_test.py]: Clarified that in the names.

**Comment by yashk2810** [comment]
Side note: maybe rename pile_map to something else? Just google `pile` and you'll see what I am talking about ;)

**Comment by axch** [comment]
@yashk2810 Yes, that's on the agenda.

**Review by mattjj — APPROVED** [review]
Woo!
